compiler/lib/js_traverse.ml

```diff
@@ -1406,29 +1401,83 @@ class simpl =
+            [ Expression_statement (EBin (Eq, v1, ECond (cond, e1, e2))), loc ]
+        (* if (e1) e2 else e3 --> if e1 ? e2 : e3 *)
+        | If_statement
+            (e1, (Expression_statement e2, _), Some (Expression_statement e3, _)) ->
+            [ Expression_statement (ECond (e1, e2, e3)), loc ]
+        (* if (!e1) e2 --> e1 || e2 *)
+        | If_statement (EUn (Not, e1), (Expression_statement e2, _), None) ->
+            [ Expression_statement (EBin (Or, e1, e2)), loc ]
+        (* if (e1) e2 --> e1 && e2 *)
+        | If_statement (e1, (Expression_statement e2, _), None) ->
+            [ Expression_statement (EBin (And, e1, e2)), loc ]
+        | Variable_statement (k, l1) ->
+            let x =
+              List.map l1 ~f:(function
+                  | DeclPattern _ as d -> Variable_statement (k, [ d ]), loc
+                  | DeclIdent (_, None) as d -> Variable_statement (k, [ d ]), loc
+                  | DeclIdent (ident, Some (exp, _)) as d -> (
+                      match assign_op (EVar ident, exp) with
+                      | Some e -> Expression_statement e, loc
+                      | None -> Variable_statement (k, [ d ]), loc))
+            in
+            x
+        | _ -> [ st, loc ]
+      in
+      (* Eliminate some expression statements. *)
+      let eliminate_expression_statements (st, loc) rem =
+        match st with
+        | Expression_statement e1 when not (is_use_strict e1) -> (
+            (* We always keep the statement `"use strict";` *)
+            match rem with
+            | [] -> [ st, loc ]
+            | (st2, loc2) :: rem -> (
+                match st2 with
+                (* e1; var x = e2; --> var x = e1, e2 *)
+                | Variable_statement (k, decls) -> (
+                    match decls with
+                    | [ DeclIdent (ident, Some (e2, e2loc)) ] ->
+                        ( Variable_statement
+                            (k, [ DeclIdent (ident, Some (ESeq (e1, e2), e2loc)) ])
+                        , loc )
+                        :: rem
+                    | _ -> (st, loc) :: (st2, loc2) :: rem)
+                (* e1; return e2; --> return e1, e2; *)
+                | Return_statement (Some e2) ->
+                    (Return_statement (Some (ESeq (e1, e2))), loc) :: rem
+                (* e1; if e2 ...; --> if (e1, e2) ...; *)
+                | If_statement (e2, iftrue, iffalse) ->
+                    (If_statement (ESeq (e1, e2), iftrue, iffalse), loc) :: rem
+                (* e1; e2; --> e1, e2; *)
+                | Expression_statement e2 ->
+                    (Expression_statement (ESeq (e1, e2)), loc) :: rem
+                | _ -> (st, loc) :: (st2, loc2) :: rem))
+        | _ -> (st, loc) :: rem
+      in
+      (* First simplify single statements *)
+      List.concat_map ~f:simplify_statement s
+      (* Then try to combine expression statements *)
+      |> List.fold_right ~f:eliminate_expression_statements ~init:[]
```

Comment:
Note that it could be nice to be able keep optimisations while having pretty printing when debugging.
Look at what's done for `shortvar` and `share` in `lib-cmdline/arg.ml`